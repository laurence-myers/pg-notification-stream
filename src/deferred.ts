import { noop } from "./utils";

interface Deferred<T> {
  promise: Promise<T>;
  reject: (error: unknown) => void;
  resolve: (value: T) => void;
}

export function deferred<T>(): Deferred<T> {
  let reject: (error: unknown) => void = noop;
  let resolve: (value: T) => void = noop;
  const promise = new Promise<T>((resolve1, reject1) => {
    resolve = resolve1;
    reject = reject1;
  });
  return {
    promise,
    reject,
    resolve,
  };
}
