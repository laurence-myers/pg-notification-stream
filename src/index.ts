export { NotificationStream } from "./notification-stream";
