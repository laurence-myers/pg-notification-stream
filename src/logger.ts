import { noop } from "./utils";

export interface Logger {
  error(message: unknown): void;
}

export const dummyLogger: Logger = {
  error: noop,
};
