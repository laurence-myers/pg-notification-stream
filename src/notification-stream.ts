import type { ClientBase } from "pg";
import { Readable, ReadableOptions } from "stream";
import { Notification } from "./notification";
import { noop } from "./utils";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Listener = (...args: any[]) => void;

function addListeners(client: ClientBase, listeners: Map<string, Listener>) {
  for (const [eventName, listener] of listeners.entries()) {
    client.addListener(eventName, listener);
  }
}

function removeListeners(client: ClientBase, listeners: Map<string, Listener>) {
  for (const [eventName, listener] of listeners.entries()) {
    client.removeListener(eventName, listener);
  }
}

enum NotificationStreamStatus {
  Pending = "Pending",
  Active = "Active",
  Stopped = "Stopped",
  Errored = "Errored",
}

export class NotificationStream extends Readable {
  protected status = NotificationStreamStatus.Pending;

  constructor(
    protected readonly client: ClientBase,
    public readonly channel: string,
    readableOptions?: ReadableOptions
  ) {
    super({
      ...readableOptions,
      objectMode: true,
    });
  }

  override _construct?(callback: (error?: Error | null) => void): void {
    try {
      const client = this.client;

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const listeners = new Map<string, (...args: any[]) => void>([
        [
          "notification",
          (message: Notification) => {
            if (message.channel === this.channel) {
              this.push(message.payload);
            }
          },
        ],
        [
          "end",
          () => {
            this.status = NotificationStreamStatus.Stopped;
            removeListeners(client, listeners);
            this.push(null);
          },
        ],
        [
          "error",
          (err: unknown) => {
            this.status = NotificationStreamStatus.Errored;
            removeListeners(client, listeners);
            this.destroy(err as Error); // lie, assume err is Error, to satisfy the interface
          },
        ],
      ]);

      addListeners(client, listeners);

      this.status = NotificationStreamStatus.Active;

      callback();
    } catch (err) {
      callback(err as Error); // lie
    }
  }

  public override _destroy(
    error: Error | null,
    callback: (error?: Error | null) => void
  ): void {
    Promise.resolve(this.client)
      .then((client): void | Promise<void> => {
        if (this.status === NotificationStreamStatus.Active) {
          return client.query(`UNLISTEN "${this.channel}";`).then(noop);
        }
      })
      .then(() => super._destroy(error, callback), callback);
  }

  override _read(): void {
    if (!this.client) {
      this.destroy(new Error(`Client not initialised`));
      return;
    }
    this.client.query(`LISTEN "${this.channel}";`).catch((err) => {
      this.destroy(err);
    });
  }
}
