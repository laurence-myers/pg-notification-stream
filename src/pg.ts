import type { Notification } from "./notification";

export interface Client {
  on(event: "error", listener: (err: Error) => void): this;
  on(event: "notification", listener: (message: Notification) => void): this;
  on(event: "end", listener: () => void): this;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  once(eventName: string | symbol, listener: (...args: any[]) => void): this;

  query<R>(queryTextOrConfig: string): Promise<R>;

  removeListener(
    eventName: string | symbol,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    listener: (...args: any[]) => void
  ): this;
}
