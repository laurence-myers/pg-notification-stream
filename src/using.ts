import type { Logger } from "./logger";

export interface Disposable {
  dispose(): Promise<void>;
}

export async function using<TResource extends Disposable, TReturn>(
  resource: TResource,
  callback: (resource: TResource) => Promise<TReturn>,
  options?: { logger: Logger }
): Promise<TReturn> {
  try {
    return await callback(resource);
  } finally {
    try {
      await resource.dispose();
    } catch (err) {
      if (options?.logger) {
        options.logger.error(err);
      }
    }
  }
}
