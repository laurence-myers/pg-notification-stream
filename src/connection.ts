import { Client } from "pg";
import { dummyLogger, Logger } from "./logger";
import type { ClientConfig } from "./pg";
import type { Disposable } from "./using";

interface PgListenIterableConnectionOptions {
  logger?: Logger;
}

export class PgListenIterableConnection implements Disposable {
  protected _client?: Client;
  protected readonly logger: Logger;

  constructor(options?: PgListenIterableConnectionOptions) {
    this.logger = options?.logger ?? dummyLogger;
  }

  public get client(): Client | undefined {
    return this._client;
  }

  async connect(config: string | ClientConfig): Promise<void> {
    if (this._client) {
      return;
    }
    const client = new Client(config);

    client.on("error", (err) => {
      this.logger.error(err);
    });

    try {
      await client.connect();
      this._client = client;
    } catch (err) {
      this.logger.error(err);
    }
  }

  public async dispose(): Promise<void> {
    if (this._client) {
      this._client.removeAllListeners();
      await this._client.end();
      this._client = undefined;
    }
  }
}

export async function connect(
  config: string | ClientConfig,
  options?: PgListenIterableConnectionOptions
): Promise<void> {
  const logger = options?.logger ?? dummyLogger;
  const client = new Client(config);

  client.on("error", (err) => {
    logger.error(err);
  });

  await client.connect();
}
