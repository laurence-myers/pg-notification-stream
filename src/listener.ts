import { Client } from "pg";
import { deferred } from "./deferred";
import { Notification } from "./notification";

export async function toAsyncIterable(
  client: Client,
  channel: string
): Promise<AsyncIterable<Notification>> {
  let nextDeferred = deferred<IteratorResult<Notification>>();

  function removeListeners() {
    for (const [eventName, listener] of listeners.entries()) {
      client.removeListener(eventName, listener);
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const listeners = new Map<string, (...args: any[]) => void>([
    [
      "notification",
      (message: Notification) => {
        nextDeferred.resolve({
          done: false,
          value: message,
        });
      },
    ],
    [
      "end",
      () => {
        removeListeners();
        nextDeferred.resolve({
          done: true,
          value: undefined,
        });
      },
    ],
    [
      "error",
      (err: unknown) => {
        removeListeners();
        nextDeferred.reject(err);
      },
    ],
  ]);

  for (const [eventName, listener] of listeners.entries()) {
    client.addListener(eventName, listener);
  }

  try {
    await client.query(`LISTEN $1;`, [channel]);
  } catch (err) {
    removeListeners();
    nextDeferred.reject(err);
    return nextDeferred.promise as Promise<unknown> as Promise<
      AsyncIterable<Notification>
    >;
  }

  return {
    [Symbol.asyncIterator]: () => ({
      async next(): Promise<IteratorResult<Notification>> {
        const value = await nextDeferred.promise;
        nextDeferred = deferred();
        return value;
      },
    }),
  };
}
