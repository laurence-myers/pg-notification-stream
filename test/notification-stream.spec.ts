import { Client } from "pg";
import { NotificationStream } from "../src/notification-stream";
import { Disposable, using } from "./using";

class DisposableClient extends Client implements Disposable {
  public dispose(): Promise<void> {
    return this.end();
  }
}

describe(NotificationStream.name, () => {
  it(`streams values as an async iterable`, async () => {
    const clientFactory = async () => {
      const client = new DisposableClient({
        connectionString:
          "postgresql://postgres:password@127.0.0.1:5432/postgres",
      });

      await client.connect();

      return client;
    };

    await using(await clientFactory(), async (notificationClient) => {
      const channel = "test_channel";

      const notificationStream = new NotificationStream(
        notificationClient,
        channel
      );

      // Start listening for messages
      const iterablePromise = (async () => {
        const notifications: string[] = [];
        for await (const notification of notificationStream) {
          notifications.push(notification);
        }
        return notifications;
      })().catch((err) => {
        expect(err).toBeUndefined();
      });

      // Send a message
      const testPayload = "Test Payload";

      await using(await clientFactory(), async (client) => {
        await client.query(`NOTIFY "${channel}", '${testPayload}';`);
      });

      // Stop the notification stream
      notificationStream.destroy();

      // Wait for the iterable to end
      const notifications = await iterablePromise;
      await notificationClient.end();

      expect(notifications).toEqual([testPayload]);
    });
  });
});
