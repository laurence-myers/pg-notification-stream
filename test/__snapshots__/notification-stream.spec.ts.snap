// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationStream async iterable API throws an error when the database connection unexpectedly disconnects: client errors 1`] = `
Array [
  [error: terminating connection due to administrator command],
  [Error: Connection terminated unexpectedly],
]
`;

exports[`NotificationStream async iterable API throws an error when the database connection unexpectedly disconnects: iterable error 1`] = `[error: terminating connection due to administrator command]`;
